app/src/main/java/org/schabi/newpipe/local/subscription/SubscriptionFragment.kt

```diff
@@ -150,28 +135,74 @@ class SubscriptionFragment : BaseStateFragment<SubscriptionState>() {
     }
 
-    private fun setupBroadcastReceiver() {
-        if (activity == null) return
+    private fun buildImportExportMenu(menu: Menu) {
+        // -- Import --
+        val importSubMenu = menu.addSubMenu(R.string.import_from)
 
-        if (subscriptionBroadcastReceiver != null) {
-            LocalBroadcastManager.getInstance(activity).unregisterReceiver(subscriptionBroadcastReceiver!!)
-        }
+        addMenuItem(importSubMenu, R.string.previous_export) { onImportPreviousSelected() }
+            .setIcon(R.drawable.ic_backup)
+
+        val services = requireContext().resources.getStringArray(R.array.service_list)
+        for (serviceName in services) {
+            try {
+                val service = NewPipe.getService(serviceName)
+
+                val subscriptionExtractor = service.subscriptionExtractor ?: continue
 
-        val filters = IntentFilter()
-        filters.addAction(EXPORT_COMPLETE_ACTION)
-        filters.addAction(IMPORT_COMPLETE_ACTION)
-        subscriptionBroadcastReceiver = object : BroadcastReceiver() {
-            override fun onReceive(context: Context, intent: Intent) {
-                _binding?.itemsList?.post {
-                    importExportItem.isExpanded = false
-                    importExportItem.notifyChanged(FeedImportExportItem.REFRESH_EXPANDED_STATUS)
+                val supportedSources = subscriptionExtractor.supportedSources
+                if (supportedSources.isEmpty()) continue
+
+                addMenuItem(importSubMenu, serviceName) {
+                    onImportFromServiceSelected(service.serviceId)
                 }
+                    .setIcon(ServiceHelper.getIcon(service.serviceId))
+            } catch (e: ExtractionException) {
+                throw RuntimeException(
+                    "Services array contains an entry that it's not a valid service name ($serviceName)",
+                    e
+                )
             }
         }
 
-        LocalBroadcastManager.getInstance(activity).registerReceiver(subscriptionBroadcastReceiver!!, filters)
+        // -- Export --
+        val exportSubMenu = menu.addSubMenu(R.string.export_to)
+
+        addMenuItem(exportSubMenu, R.string.file) { onExportSelected() }
+            .setIcon(R.drawable.ic_save)
+    }
+
+    private fun addMenuItem(
+        subMenu: SubMenu,
+        @StringRes title: Int,
+        onClick: Runnable
+    ): MenuItem {
+        return addMenuItem({ subMenu.add(title) }, onClick)
+    }
+
+    private fun addMenuItem(
+        subMenu: SubMenu,
+        title: String,
+        onClick: Runnable
+    ): MenuItem {
+        return addMenuItem({ subMenu.add(title) }, onClick)
+    }
+
+    private fun addMenuItem(
+        menuItemSupplier: Supplier<MenuItem>,
+        onClick: Runnable
+    ): MenuItem {
+        val item = menuItemSupplier.get()
+
+        item.setOnMenuItemClickListener { _ ->
+            onClick.run()
+            true
+        }
+
+        return item
     }
```

Comment:
No need for `Supplier`s and no need for return values ;-)
```suggestion
    private fun addMenuItemToSubmenu(
        subMenu: SubMenu,
        @StringRes title: Int,
        onClick: Runnable
    ) {
        setClickListenerToMenuItem(subMenu.add(title), onClick)
    }

    private fun addMenuItemToSubmenu(
        subMenu: SubMenu,
        title: String,
        onClick: Runnable
    ) {
        setClickListenerToMenuItem(subMenu.add(title), onClick)
    }

    private fun addMenuItem(
        menuItem: MenuItem,
        onClick: Runnable
    ) {
        menuItem.setOnMenuItemClickListener { _ ->
            onClick.run()
            true
        }
    }
```

Reply:
Fixed